src/python/pants/rules/core/strip_source_roots.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import cast
+
+from pants.build_graph.files import Files
+from pants.engine.fs import EMPTY_SNAPSHOT, Digest, DirectoryWithPrefixToStrip, Snapshot
+from pants.engine.legacy.graph import HydratedTarget
+from pants.engine.rules import rule, subsystem_rule
+from pants.engine.selectors import Get
+from pants.source.source_root import NoSourceRootError, SourceRootConfig
+
+
+@dataclass(frozen=True)
+class SourceRootStrippedSources:
+  """Wrapper for a snapshot of targets whose source roots have been stripped."""
+  snapshot: Snapshot
+
+
+@dataclass(frozen=True)
+class StripSourceRootsRequest:
+  """A request to strip source roots for every file in the snapshot.
+
+  The field `sentinel_file` is used to determine the source root for the files to be stripped. This
+  assumes that every file shares the same source root, which should be true in practice as the
+  `sources` field for a target always has files in the same source root (outside of aggregate
```

Comment:
Not sure this discussion of "aggregate targets" is meaningful. Those (as described through that link, but under the name "target aggregates") don't have `sources` fields at all, they exist just to provide deps.